blueocean-github-pipeline/src/main/java/io/jenkins/blueocean/blueocean_github_pipeline/GithubEnterpriseScm.java

```diff
@@ -30,14 +35,46 @@ public GithubEnterpriseScm(Reachable parent) {
     }
 
     @Override
+    public @Nonnull String getUri() {
+        String apiUri = getCustomApiUri();
+
+        // NOTE: GithubEnterpriseScm requires that the apiUri be specified
+        if (StringUtils.isEmpty(apiUri)) {
+            throw new ServiceException.BadRequestException(new ErrorMessage(400, "URI is required"));
+        }
+
+        return apiUri;
+    }
+
+    @Override
+    public String getCredentialId() {
+        // there is no "default" credential for GitHub Enterprise, so return null
+        return null;
+    }
+
+    @Override
     public String getCredentialDomainName() {
-        java.net.URI uri;
-        try {
-            uri = new URI(getUri());
-        } catch (URISyntaxException e) {
-            throw new ServiceException.UnexpectedErrorException(new ErrorMessage(400, "Invalid Github Enterprise URI: "+getUri()));
+        return DOMAIN_NAME;
+    }
+
+    @Exported(name = CREDENTIALS)
+    public List<GithubEnterpriseScmCredential> getCredentials() {
```

Comment:
@vivek So are you proposing for GHE that we call general-purpose "credential search" API similar to how we do today for Git pipelines? Would we make similar changes to existing GH code too? In principle that sounds fine to me. However, we need to verify we can still correlate the `credentialId` to the `apiUrl` the user specified. In the current credentials model I don't think this is possible?

If I am misunderstanding your suggested approach then perhaps we can jump on a call in a couple hours if you are available.